syntax = "proto3";
import "google/protobuf/any.proto";
import "SerializableExpression.proto";
package NES;


/**
The serializable wrapper for a schema object
 */
message SerializableSchema{
  repeated SerializableField fields = 1;
}

/**
The serializable wrapper for the nes partition object
 */
message NesPartition {
  uint64 queryId = 1;
  uint64 operatorId = 2;
  uint64 partitionId = 3;
  uint64 subpartitionId = 4;
}

/**
The serializable wrapper for the node location object
 */
message NodeLocation {
  uint64 nodeId = 1;
  string hostname = 2;
  uint32 port = 3;
}

/*
The serializable wrapper definition for operator nodes
 */
message SerializableOperator{

  /**
  Source Details contains properties for the source operator, such as the source descriptor
   */
  message SourceDetails {
    message SerializableBinarySourceDescriptor {
      string filePath = 1;
      SerializableSchema sourceSchema = 2;
    }

    message SerializableCsvSourceDescriptor {
      string filePath = 1;
      string delimiter = 2;
      uint64 numberOfTuplesToProducePerBuffer = 3;
      uint64 numBuffersToProcess = 4;
      uint64 frequency = 5;
      SerializableSchema sourceSchema = 6;
    }

    message SerializableDefaultSourceDescriptor {
      uint64 numBuffersToProcess = 1;
      uint64 frequency = 2;
      SerializableSchema sourceSchema = 3;
    }

    message SerializableKafkaSourceDescriptor {
      string brokers = 1;
      string topic = 2;
      string groupId = 3;
      bool autoCommit = 4;
      uint64 kafkaConnectTimeout = 5;
      SerializableSchema sourceSchema = 6;
    }

    message SerializableLogicalStreamSourceDescriptor {
      string streamName = 1;
      SerializableSchema sourceSchema = 2;
    }

    message SerializableSenseSourceDescriptor {
      string udfs = 1;
      SerializableSchema sourceSchema = 2;
    }

    message SerializableZMQSourceDescriptor {
      string host = 1;
      uint32 port = 2;
      SerializableSchema sourceSchema = 3;
    }

    message SerializableOPCSourceDescriptor {
      string url = 1;
      uint32 namespaceIndex = 2;
      string identifier = 3;
      uint32 identifierType = 4;
      string user = 5;
      string password = 6;
      SerializableSchema sourceSchema = 7;
    }

    message SerializableNetworkSourceDescriptor {
      SerializableSchema sourceSchema = 1;
      NesPartition nesPartition = 2;
    }

    google.protobuf.Any sourceDescriptor = 1;
  }

  /**
  Sink Details contains properties for the sink operator, such as the sink descriptor
  */
  message SinkDetails {
    message SerializableFileSinkDescriptor {
      string filePath = 1;
      bool append = 2;
      string sinkFormat = 3;
    }

    message SerializableKafkaSinkDescriptor {
      string topic = 1;
      string brokers = 2;
      uint64 kafkaConnectTimeout = 5;
    }

    message SerializablePrintSinkDescriptor {
    }

    message SerializableZMQSinkDescriptor {
      string host = 1;
      uint32 port = 2;
      bool isInternal = 3;
    }

    message SerializableOPCSinkDescriptor {
      string url = 1;
      uint32 namespaceIndex = 2;
      string identifier = 3;
      uint32 identifierType = 4;
      string user = 5;
      string password = 6;
    }

    message SerializableNetworkSinkDescriptor {
      NodeLocation nodeLocation = 1;
      NesPartition nesPartition = 2;
      uint64 waitTime = 3;
      uint32 retryTimes = 4;
    }

    google.protobuf.Any sinkDescriptor = 1;
  }

  /**
   Map Details contains properties for the map operator, such as the field assignment expression
   */
  message MapDetails {
    SerializableExpression expression = 1;
  }

  /**
   Merge Details contains properties for the merge operator, which is empty currently.
   */
  message MergeDetails {}

  /**
   Broadcast details contains properties for the broadcast operator
   */
  message BroadcastDetails{}

  /**
   Filter Details contains properties for the filter operator, such as the filter predicate expression
   */
  message FilterDetails {
    SerializableExpression predicate = 1;
  }

  /**
    Window Details contains properties for the window operator
  */
  message WindowDetails {

    message DistributionCharacteristic {
      enum Distribution{
        Complete = 0;
        Slicing = 1;
        Combining = 3;
      }
      Distribution distr = 1;
    }

    message TimeCharacteristic {
      enum Type{
        EventTime = 0;
        ProcessingTime = 1;
      }
      Type type = 1;
      string field = 2;
    }
    message TumblingWindow {
      TimeCharacteristic timeCharacteristic = 1;
      uint32 size = 2;
    }
    message SlidingWindow {
      TimeCharacteristic timeCharacteristic = 1;
      uint32 size = 2;
      uint32 slide = 3;
    }
    message Aggregation {
      enum Type{
        SUM = 0;
        MAX = 1;
        MIN = 2;
        COUNT = 3;
      }
      Type type = 1;
      SerializableExpression onField = 2;
      SerializableExpression asField = 3;
    }
    google.protobuf.Any windowType = 1;
    Aggregation windowAggregation = 2;
    DistributionCharacteristic distrChar = 3;
    SerializableExpression onKey = 4;
    uint64 numberOfInputEdges = 5;

  }
  google.protobuf.Any details = 1;
  repeated SerializableOperator children = 2;
  SerializableSchema inputSchema = 3;
  SerializableSchema outputSchema = 4;
  uint64 operatorId = 5;
}